xamarin/java.interop
# [build] Remove globalPackagesFolder override

Context: dec2e3902514cafad004203ca783ca87ee38ed49

How does one run NUnit unit-tests?  `dotnet test` is an answer --
and perhaps we should consider fully migrating to it -- but for now
we currently use `nunit3-console.exe`.

How do we find the path to `nunit3-console.exe`?  *This* is why we
set the [`globalPackagesFolder`][0] property in `NuGet.config` in
commit dec2e390: `globalPackagesFolder` tells NuGet where to place
downloaded NuGet packages.  By setting `globalPackagesFolder`, we
knew where the `NUnit.ConsoleRunner` artifacts were installed, and
could construct a path to `nunit3-console.exe`.

The problem with `globalPackagesFolder` is that it's duplication:
the `packages` folder is our own special copy of our NuGet packages,
unshared by anything else on the system.  This increases disk space
requirements, and slows down provisioning (as those artifacts need
to be re-downloaded whenever the repo is `git clean`ed).

There is an alternative: when the
[`%(PackageReference.GeneratePathProperty)`][1] property is True,
a `$(Pkg[NuGet Package Name])` property is created, which contains
the path to where the NuGet package is installed.  Thus, instead of
needing to hardcode:

	packages\nunit.consolerunner\3.12.0\tools\nunit3-console.exe

we can instead use:

	$(PkgNUnit_ConsoleRunner)/tools/nunit3-console.exe

…in theory.

In practice, `$(PkgNUnit_ConsoleRunner)` isn't set until after
`@(PackageReference)` is processed and (lots of) other MSBuild
targets are run.  Targets which *aren't* run when our `RunTests`
MSBuild target is run.

Allow `RunNUnitTests.targets` to use `$(PkgNUnit_ConsoleRunner)`
by adding a new `build-tools\PackagePaths\PackagePaths.csproj`
project, which will create
`bin/Build$(Configuration)/PackagePaths.props` as part of its build.
`RunNUnitTests.targets` can then import `PackagePaths.props`,
obtaining access to `$(PkgNUnit_ConsoleRunner)`.

A "philosophical" problem arises, around *how to build*
`PackagePaths.csproj` from `make prepare`.  We effectively have two
separate and overlapping build systems: the original `make` system,
and a `dotnet build`-based system:

	% make prepare
	% make all
	% make run-all-tests

	% make prepare-core
	% dotnet build -t:Prepare
	% dotnet build
	# run various tests via YAML…

*Begin* unifying these systems, by updating `make prepare` to run
`msbuild -target:Prepare Java.Interop.sln`, and removing make targets
which overlap with the `Prepare` MSBuild target.  This allows
`make prepare` to create `PackagePaths.props`, and reduces the amount
of Makefile code that we need to maintain.

Quasi-related:

Rename the `<GenerateVersionFile/>` task to `<ReplaceFileContents/>`.
This better matches what the task actually does.

Remove `jdk.mk` and `jdk.targets`.  These are no longer needed.

Move `Java.Interop.BootstrapTasks.sln` into topdir.  This is so that
it can share `global.json` settings.

[0]: https://docs.microsoft.com/en-us/nuget/Consume-Packages/managing-the-global-packages-and-cache-folders
[1]: https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#generatepathproperty
